src_kt/java/arcs/storage/api/ArcsSingleton.kt

```diff
@@ -0,0 +1,337 @@
+            }
+        },
+        coroutineContext = coroutineContext
+    )
+}
+
+/**
+ * Creates an [ArcsSingleton] which manages a reference to a [ReferencablePrimitive] object of
+ * type <T> located at the given [storageKey].
+ *
+ * **Note:** By supplying your [coroutineContext], any bindings to the storage layer are released
+ * when the context's primary job is completed, thus avoiding a memory leak.
+ */
+@Suppress("FunctionName")
+@ExperimentalCoroutinesApi
+inline fun <reified T> ArcsSingleton(
+    storageKey: StorageKey,
+    existenceCriteria: ExistenceCriteria = ExistenceCriteria.MayExist,
+    coroutineContext: CoroutineContext = Dispatchers.Default
+): ArcsSingleton<ReferencablePrimitive<T>,
+    CrdtSingleton.Data<ReferencablePrimitive<T>>,
+    CrdtSingleton.IOperation<ReferencablePrimitive<T>>> {
+
+    require(ReferencablePrimitive.isSupportedPrimitive(T::class)) {
+        "Unsupported type: ${T::class}"
+    }
+
+    val storeOps = StoreOptions<
+        CrdtSingleton.Data<ReferencablePrimitive<T>>,
+        CrdtSingleton.IOperation<ReferencablePrimitive<T>>,
+        ReferencablePrimitive<T>>(
+        storageKey = storageKey,
+        // TODO: this is wrong. There should probably be a PrimitiveType?
+        type = SingletonType(CountType()),
+        existenceCriteria = existenceCriteria,
+        mode = StorageMode.Direct
+    )
+    return ArcsSingleton(
+        store = Store(storeOps),
+        toStoreData = { it },
+        toStoreOp = { it },
+        coroutineContext = coroutineContext
+    )
+}
+
+/**
+ * [ArcsSingleton] is a register-like data structure which allows for easy interaction with
+ * [CrdtSingleton]s managed by Arcs' storage layer.
+ *
+ * All functions on [ArcsSingleton] are suspending functions, due to the fact that communicating
+ * with the storage layer is an inherently asynchronous process.
+ *
+ * **Note:** By supplying your [CoroutineContext] to the constructor, any bindings to the storage
+ * layer are released when the context's primary job is completed, thus avoiding a memory leak.
+ */
+class ArcsSingleton<T, StoreData, StoreOp>(
+    private val store: Store<StoreData, StoreOp, T>,
+    private val toStoreData: (CrdtSingleton.Data<T>) -> StoreData,
+    private val toStoreOp: (CrdtSingleton.IOperation<T>) -> StoreOp,
+    coroutineContext: CoroutineContext
+) : Referencable
+    where T : Referencable,
+          StoreData : CrdtSingleton.Data<T>,
+          StoreOp : CrdtSingleton.IOperation<T> {
+
+    override val id: ReferenceId = store.storageKey.toKeyString()
+
+    /**
+     * The [Actor] this instance will use when performing Crdt operations on the underlying data.
+     */
+    val actor: Actor by lazy { "ArcsSingleton@${hashCode()}" }
+
+    private val scope = CoroutineScope(coroutineContext)
+    private val crdtMutex = Mutex()
+    private val crdtSingleton = CrdtSingleton<T>()//by guardWith(crdtMutex, CrdtSingleton<T>())
```

Comment:
Delete comment?

Reply:
Whoops, forgot to uncomment this, actually.